File: day02/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type passwordInfo struct {
	first    int
	second   int
	target   string
	password string
}

func partTwo(passwords []passwordInfo) (valid int) {
	for _, password := range passwords {

		if (string(password.password[password.first]) == password.target) !=
			(string(password.password[password.second]) == password.target) {
			valid++
		}
	}
	return
}

func partOne(passwords []passwordInfo) (valid int) {
	for _, password := range passwords {
		count := strings.Count(password.password, password.target)

		if password.first <= count && count <= password.second {
			valid++
		}
	}
	return
}

func main() {
	reader, _ := os.Open("input.txt")
	scanner := bufio.NewScanner(reader)
	passwords := make([]passwordInfo, 0)
	for scanner.Scan() {
		line := scanner.Text()
		splitLine := strings.Split(line, ":")
		policy, password := splitLine[0], splitLine[1]

		policySplit := strings.Split(policy, " ")
		target := policySplit[1]
		letterRange := strings.Split(policySplit[0], "-")
		first, _ := strconv.Atoi(letterRange[0])
		second, _ := strconv.Atoi(letterRange[1])

		if password != "" {
			passwords = append(passwords, passwordInfo{first: first, second: second, target: target, password: password})
		}
	}

	fmt.Println(partOne(passwords))
	fmt.Println(partTwo(passwords))
}

File: day02/solution.cpp
#include <iostream>
#include <fstream>
#include <tuple>
#include <vector>

using namespace std;

typedef tuple<int, int, char, string> Password;

int part1(const vector<Password> &passwords) {
    int valid = 0;
    for (auto pass_tuple : passwords) {
        int low, high;
        char target;
        string password;

        tie(low, high, target, password) = pass_tuple;

        int count = 0;
        for (char c : password) {
            if (c == target) {
                count++;
            }
        }

        if (low <= count && count <= high) {
            valid++;
        }
    }
    return valid;
}

int part2(const vector<Password> &passwords) {
    int valid = 0;
    for (auto pass_tuple : passwords) {
        int first, second;
        char target;
        string password;

        tie(first, second, target, password) = pass_tuple;

        if ((password[first - 1] == target) != (password[second - 1] == target)) {
            valid++;
        }
    }
    return valid;
}

int main() {
    ifstream in("input.txt");

    vector<Password> passwords;
    while (!in.eof()) {
        int num1, num2;
        char dash, colon;
        char character;
        string password;

        in >> num1 >> dash >> num2 >> character >> colon >> password;
        if (!password.empty()) {
            passwords.push_back(make_tuple(num1, num2, character, password));
        }
    }

    cout << part1(passwords) << endl << part2(passwords) << endl;
}